editor/gui/scene_tree_editor.cpp

```diff
@@ -1548,16 +1572,32 @@ void SceneTreeDialog::set_valid_types(const Vector<StringName> &p_valid) {
 		HBoxContainer *hb = memnew(HBoxContainer);
 		hflow->add_child(hb);
 
+		// Attempt to get the correct name and icon for script path types.
+		String name = type;
+		Ref<Texture2D> icon = EditorNode::get_singleton()->get_class_icon(type);
+
+		// If we can't find a global class icon, try to find one for the script.
+		if (icon.is_null() && ResourceLoader::exists(type, "Script")) {
+			Ref<Script> node_script = ResourceLoader::load(type);
+			if (node_script.is_valid()) {
+				name = name.get_file();
+				icon = EditorNode::get_editor_data().get_script_icon(node_script);
+				if (icon.is_null()) {
+					icon = EditorNode::get_singleton()->get_class_icon(node_script->get_class_name());
+				}
```

Comment:
You should use `EditorNode::get_singleton()->get_object_icon(node_script)` which will handle all of this for you.

Reply:
Ah thanks! I missed the fact that this function has this line:
`if (scr.is_null() && p_object->is_class("Script")) {`
I thought it only really worked if you had a node.